File: frontend/src/lib/api/video.ts
/**
 * 视频生成 API 客户端
 */

const API_BASE_URL = import.meta.env.VITE_API_URL || '/api'

export interface CreateVideoTaskParams {
  prompt: string
  model?: string
  aspect_ratio?: string
  seconds?: number
  image?: string
  /** 额外参数，与 prompt 同级平铺到请求体 */
  extraParams?: Record<string, unknown>
}

/** 查询任务响应：进行中 / 完成 / 失败，见 API 文档 */
export interface VideoTaskResponse {
  id?: string
  task_id?: string
  /** queued | in_progress | completed | failed */
  status?: string
  created_at?: string
  completed_at?: string
  model?: string
  /** 仅完成时返回，有效期 1 小时 */
  video_url?: string
  usage?: {
    seconds?: number
    video_count?: number
    size?: string
  }
  /** 仅失败时返回 */
  error?: string
  raw_response?: Record<string, unknown>
  [key: string]: unknown
}

/**
 * 提交视频生成任务（文生视频或图生视频）
 */
export async function createVideoTask(
  params: CreateVideoTaskParams
): Promise<VideoTaskResponse> {
  const body: Record<string, unknown> = {
    prompt: params.prompt,
    aspect_ratio: params.aspect_ratio ?? '16:9',
    seconds: params.seconds ?? 5
  }
  if (params.model != null) body.model = params.model
  if (params.image != null) body.image = params.image
  if (params.extraParams && Object.keys(params.extraParams).length > 0) {
    Object.assign(body, params.extraParams)
  }

  const response = await fetch(`${API_BASE_URL}/videos`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  })
  if (!response.ok) {
    const err = await response.json().catch(() => ({}))
    throw new Error(
      (err as { detail?: string })?.detail
        ? String((err as { detail: string }).detail)
        : '提交视频任务失败'
    )
  }
  return response.json()
}

/**
 * 查询视频任务结果（轮询用）
 * @param taskId 创建任务时返回的 id
 * @param provider 可选，供应商名称（如 火山方舟），任务不在本地缓存时指定可加快查询
 */
export async function getVideoTaskResult(
  taskId: string,
  provider?: string
): Promise<VideoTaskResponse> {
  const url = new URL(`${API_BASE_URL}/videos/${taskId}`)
  if (provider) url.searchParams.set('provider', provider)
  const response = await fetch(url.toString(), { cache: 'no-store' })
  if (!response.ok) {
    const err = await response.json().catch(() => ({}))
    throw new Error(
      (err as { detail?: string })?.detail
        ? String((err as { detail: string }).detail)
        : '查询视频任务失败'
    )
  }
  return response.json()
}

File: backend/app/api/routes/video.py
import json
import logging

import aiohttp

from fastapi import APIRouter, HTTPException

from ...config import settings
from ...schemas.video import VideoCreateRequest

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/videos", tags=["Video"])


def _headers() -> dict[str, str]:
    return {
        "Authorization": f"Bearer {settings.openai_api_key}",
        "Content-Type": "application/json",
    }


@router.post("")
async def create_video_task(request: VideoCreateRequest):
    """
    提交视频生成任务，请求即梦视频生成 API。
    返回任务 id、status、model 等，可用于后续轮询结果。
    """
    if not settings.openai_api_key:
        raise HTTPException(
            status_code=503,
            detail="未配置 OPENAI_API_KEY，无法调用视频生成服务。",
        )
    url = settings.video_api_url
    payload = {
        "model": request.model or settings.video_model,
        "prompt": request.prompt,
        "aspect_ratio": request.aspect_ratio,
        "seconds": request.seconds,
    }
    if request.image is not None:
        payload["image"] = request.image
    logger.info("给 aiping 发送请求: url=%s body=%s", url, payload)
    timeout = aiohttp.ClientTimeout(total=settings.openai_timeout)
    try:
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.post(url, headers=_headers(), json=payload) as response:
                text = await response.text()
                logger.info("收到 aiping 响应: status=%s body=%s", response.status, text or "(empty)")
                if response.content_type and "application/json" in response.content_type and text:
                    try:
                        result = json.loads(text)
                    except json.JSONDecodeError:
                        result = {"raw": text}
                else:
                    result = {"raw": text} if text else {}
                if response.status >= 400:
                    logger.warning("视频任务提交失败: status=%s body=%s", response.status, text)
                    raise HTTPException(
                        status_code=response.status,
                        detail=result if isinstance(result, dict) else {"message": text},
                    )
                return result
    except aiohttp.ClientError as e:
        logger.exception("请求视频 API 失败: %s", e)
        raise HTTPException(status_code=502, detail=f"请求视频服务失败: {e}") from e


@router.get("/{task_id}")
async def get_video_task_result(task_id: str, provider: str | None = None):
    """
    根据任务 ID 查询视频生成结果。
    状态：queued / in_progress / completed / failed。
    完成时响应含 video_url；失败时含 error。
    可选 Query：provider（如 火山方舟）加快查询。
    """
    if not settings.openai_api_key:
        raise HTTPException(
            status_code=503,
            detail="未配置 OPENAI_API_KEY，无法调用视频生成服务。",
        )
    url = f"{settings.video_api_url}/{task_id}"
    params = {"provider": provider} if provider else None
    logger.info("给 aiping 发送请求: url=%s params=%s", url, params)
    timeout = aiohttp.ClientTimeout(total=settings.openai_timeout)
    try:
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.get(url, headers=_headers(), params=params) as response:
                text = await response.text()
                logger.info("收到 aiping 响应: status=%s body=%s", response.status, text or "(empty)")
                if response.content_type and "application/json" in response.content_type and text:
                    try:
                        result = json.loads(text)
                    except json.JSONDecodeError:
                        result = {"raw": text}
                else:
                    result = {"raw": text} if text else {}
                if response.status >= 400:
                    logger.warning("查询视频任务失败: task_id=%s status=%s body=%s", task_id, response.status, text)
                    raise HTTPException(
                        status_code=response.status,
                        detail=result if isinstance(result, dict) else {"message": text},
                    )
                return result
    except aiohttp.ClientError as e:
        logger.exception("请求视频 API 失败: %s", e)
        raise HTTPException(status_code=502, detail=f"请求视频服务失败: {e}") from e
